modules/reconciliationRtdProvider.js

```diff
@@ -0,0 +1,313 @@
+            ? adDeliveryId[0]
+            : utils.generateUUID();
+        }
+      }
+    }
+
+    // Call local impression callback
+    const args = Object.assign({}, data.args, {
+      publisherDomain: window.location.hostname,
+      publisherMemberId: _moduleParams.publisherMemberId,
+      adUnitId,
+      adDeliveryId,
+    });
+
+    utils.triggerPixel(`${_moduleParams.impressionUrl}?${stringify(args)}`);
+
+    // Send response back to the Advertiser tag
+    let response = {
+      type: MessageType.IMPRESSION_RESPONSE,
+      id: data.id,
+      args: Object.assign(
+        {
+          publisherDomain: window.location.hostname,
+        },
+        data.args
+      ),
+    };
+
+    // If access is allowed - add ad unit id to response
+    if (_moduleParams.allowAccess) {
+      Object.assign(response.args, {
+        adUnitId,
+        adDeliveryId,
+      });
+    }
+
+    e.source.postMessage(JSON.stringify(response), '*');
+  }
+}
+
+/**
+ * Get top iframe window for nested Window object
+ * - top
+ * -- iframe.window  <-- top iframe window
+ * --- iframe.window
+ * ---- iframe.window <-- win
+ *
+ * @param {Window} win nested iframe window object
+ */
+export function getTopIFrameWin(win) {
+  if (!win) {
+    return null;
+  }
+
+  try {
+    while (win.parent !== win.top) {
+      win = win.parent;
+    }
+    return win;
+  } catch (e) {
+    return null;
+  }
+}
+
+/**
+ * get all slots on page
+ * @return {Object[]} slot GoogleTag slots
+ */
+function getAllSlots() {
+  return utils.isGptPubadsDefined() && window.googletag.pubads().getSlots();
+}
+
+/**
+ * get GPT slot by placement id
+ * @param {string} code placement id
+ * @return {?Object}
+ */
+function getSlotByCode(code) {
+  const slots = getAllSlots();
+  if (!slots || !slots.length) {
+    return null;
+  }
+  return (
+    find(
+      slots,
+      (s) => s.getSlotElementId() === code || s.getAdUnitPath() === code
+    ) || null
+  );
+}
+
+/**
+ * get GPT slot by iframe window
+ * @param {Window} win
+ * @return {?Object}
+ */
+function getSlotByWin(win) {
+  const slots = getAllSlots();
+
+  if (!slots || !slots.length) {
+    return null;
+  }
+
+  return (
+    find(slots, (s) => {
+      let slotElement = document.getElementById(s.getSlotElementId());
+
+      if (slotElement) {
+        let slotIframe = slotElement.querySelector('iframe');
+
+        if (slotIframe && slotIframe.contentWindow === win) {
+          return true;
+        }
+      }
+
+      return false;
+    }) || null
+  );
+}
+
+/**
+ * serialize object and return query params string
+ * @param {Object} data
+ * @return {string}
+ */
+export function stringify(query) {
+  const parts = [];
+
+  for (let key in query) {
+    if (query.hasOwnProperty(key)) {
+      let val = query[key];
+      if (typeof query[key] !== 'object') {
+        parts.push(`${key}=${encodeURIComponent(val)}`);
+      } else {
+        parts.push(`${key}=${encodeURIComponent(stringify(val))}`);
```

Comment:
Since this is recursively calling `stringify`, I think the result could end up looking like:
```
key=key=value
``` 
Is this the intended behavior?

Reply:
Yes, it is supposed to work this way: nested objects are stringified to query-string and URL-encoded:
`{foo: 'bar', nested: {key1: 'value1', key2: 'value2'}}` 
becomes
`foo=bar&nested=key1%3Dvalue1%26key2%3Dvalue2`